src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/DoNotUseReferenceEqualsWithValueTypesTests.cs

```diff
@@ -0,0 +1,246 @@
+// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Testing;
+using Xunit;
+using VerifyCS = Test.Utilities.CSharpCodeFixVerifier<
+    Microsoft.NetCore.Analyzers.Runtime.DoNotUseReferenceEqualsWithValueTypesAnalyzer,
+    Microsoft.CodeAnalysis.Testing.EmptyCodeFixProvider>;
+using VerifyVB = Test.Utilities.VisualBasicCodeFixVerifier<
+    Microsoft.NetCore.Analyzers.Runtime.DoNotUseReferenceEqualsWithValueTypesAnalyzer,
+    Microsoft.CodeAnalysis.Testing.EmptyCodeFixProvider>;
+
+namespace Microsoft.NetCore.Analyzers.Runtime.UnitTests
+{
+    public class DoNotUseReferenceEqualsWithValueTypesTests
+    {
+        [Fact]
+        public async Task ReferenceTypesAreOK()
+        {
+            await VerifyCS.VerifyAnalyzerAsync(@"
+using System;
+
+namespace TestNamespace
+{
+    class TestClass
+    {
+        private static bool TestMethod(string test)
+        {
+            return ReferenceEquals(test, string.Empty);
+        }
+    }
+}");
+
+            await VerifyVB.VerifyAnalyzerAsync(@"
+Imports System
+
+Namespace TestNamespace
+    Class TestClass
+        Private Shared Function TestMethod(test as String)
+            Return ReferenceEquals(string.Empty, test)
+        End Function
+    End Class
+End Namespace");
+        }
+
+        [Fact]
+        public async Task LeftArgumentFailsForValueType()
+        {
+            await VerifyCS.VerifyAnalyzerAsync(@"
+using System;
+
+namespace TestNamespace
+{
+    class TestClass
+    {
+        private static bool TestMethod(string test)
+        {
+            return ReferenceEquals(IntPtr.Zero, test);
+        }
+    }
+}",
+                GetCSharpResultAt(10, 36, "objA", "System.IntPtr"));
```

Comment:
I believe the test framework also supports annotating the source such as `return ReferenceEquals([|IntPtr.Zero|], test);`, so you can avoid passing in this explicit expected diagnostic and the framework will decode the annotation and validate the diagnostic. I am fine with your current form as well, if that is by preference.